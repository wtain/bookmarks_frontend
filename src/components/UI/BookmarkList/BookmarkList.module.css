
.bookmark {
    display: flex;
}
.bookmarkEnter {
    /* transform: translateX(-350px); */
    opacity: 0;
}
.bookmarkEnterActive {
    /* transform: translateX(0px); */
    opacity: 1;
    transition: opawicity 700ms;
}
.bookmarkExit {
    opacity: 1;
}
.bookmarkExitActive {
    /* transform: translateX(-350px); */
    opacity: 0;
    transition: opacity 300ms ease-out;
}