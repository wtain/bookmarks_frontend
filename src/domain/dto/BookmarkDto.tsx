
import React from "react";
import TagDto from "./TagDto";

interface BookmarkDto {
    summary: string;
    contents: string;
    created: string;
    updated?: string;
    id: string;
    tags: TagDto[];
    isDone: boolean;
}

export default BookmarkDto;