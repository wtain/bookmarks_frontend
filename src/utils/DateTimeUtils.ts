import moment from "moment";

export function parseDateTimeFromMongo(datetime: Date | undefined) {
  if (datetime === undefined) {
    return undefined;
  }
  return moment(datetime.toISOString()).fromNow();
}

export function addOneMonth(date: Date): Date {
  let result = new Date(date);
  result.setMonth(result.getMonth() + 1)
  return result;
}

export function firstDayOfMonth(date: Date): Date {
  let result = new Date(date);
  result.setDate(1)
  return result;
}

export function getDatePortion(date: Date): string {
  const dateCopy = new Date(date);
  dateCopy.setUTCHours(0);
  dateCopy.setMinutes(0);
  const momentForDate = moment(dateCopy);
  return momentForDate.format('YYYY-MM-DD');
}

export function datesEqual(date1: Date, date2: Date): boolean {
  return getDatePortion(date1) === getDatePortion(date2);
}