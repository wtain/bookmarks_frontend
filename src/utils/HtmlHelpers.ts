import sanitizeHtml from "sanitize-html";

export function sanitizeText(text: string){
  return sanitizeHtml(text, {
    allowedTags: ['br', 'b', 'a', 'img']
  });
}