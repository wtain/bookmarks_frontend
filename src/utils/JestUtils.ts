
export function jestHacks() {
    // https://stackoverflow.com/questions/39830580/jest-test-fails-typeerror-window-matchmedia-is-not-a-function
    // https://github.com/facebook/create-react-app/issues/10126
    window.matchMedia = (query) => ({
        matches: false,
        media: query,
        onchange: null,
        addListener: jest.fn(), // Deprecated
        removeListener: jest.fn(), // Deprecated
        addEventListener: jest.fn(),
        removeEventListener: jest.fn(),
        dispatchEvent: jest.fn(),
      })
}