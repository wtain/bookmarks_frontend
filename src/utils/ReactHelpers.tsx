
import React, { MutableRefObject } from "react";

const scrollToRef = (ref: MutableRefObject<null | HTMLDivElement>) => {
    if (ref.current) {
        ref.current!.scrollIntoView({
            behavior: "smooth",
          });
        // On mobile Chrome smooth is not supported, so let's 
        // make sure we're scrolled anyway
        // ref.current!.scrollIntoView(true);
    }
};

const delay = (milliseconds: number) => {
    return new Promise(resolve => setTimeout(resolve, milliseconds));
}

const ReactHelpers = {
    scrollToRef, delay
}

export default ReactHelpers;