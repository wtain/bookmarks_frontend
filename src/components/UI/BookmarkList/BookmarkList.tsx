
import React from "react";
import BookmarkDto from "../../../domain/dto/BookmarkDto";
import Bookmark from "../Bookmark/Bookmark";
import {TransitionGroup, CSSTransition} from "react-transition-group";
import cl from './BookmarkList.module.css'
import { Flex } from "reflexy";
import TagDto from "../../../domain/dto/TagDto";

interface Props {
    bookmarks: BookmarkDto[];
    onRemoveBookmark: (bm: BookmarkDto) => void;
    onBookmarkContentsChanged: (bm: BookmarkDto, new_contents: string) => void;
    onBookmarkSummaryChanged: (bm: BookmarkDto, new_summary: string) => void;
    onBookmarkTagAdded: (bm: BookmarkDto, new_tag: TagDto) => void;
    onBookmarkTagRemoved: (bm: BookmarkDto, index: number) => void;
    onBookmarkIsDoneChanged: (bm: BookmarkDto, new_value: boolean) => void;
    newBookmarkId? :string;
}

const BookmarkList: React.FC<Props> = (props: Props) => {
    if (props.bookmarks.length === 0) {
        return (<div>(No bookmarks)</div>)
    }
    return (
        <div>
            <Flex>
                <TransitionGroup>
                {
                    props.bookmarks.map(bm => 
                            <CSSTransition 
                                key={bm.id} 
                                timeout={500} 
                                classNames={{
                                    enterActive: cl.bookmarkEnterActive,
                                    enter: cl.bookmarkEnter,
                                    exit: cl.bookmarkExit,
                                    exitActive: cl.bookmarkExitActive
                                }}>
                                <Bookmark key={bm.id}
                                        highlightJustAdded={props.newBookmarkId! === bm.id}
                                        bookmark={bm} 
                                        doRemove={props.onRemoveBookmark}
                                        doChangeContents={(new_contents: string) => {
                                            props.onBookmarkContentsChanged(bm, new_contents)
                                        }}
                                        doChangeSummary={(new_summary: string) => {
                                            props.onBookmarkSummaryChanged(bm, new_summary)
                                        }}
                                        doAddTag={(tag: TagDto) => {
                                            props.onBookmarkTagAdded(bm, tag);
                                        }}
                                        doRemoveTag={(index: number) => {
                                            props.onBookmarkTagRemoved(bm, index);
                                        }}
                                        doChangeIsDone={(new_value: boolean) => {
                                            props.onBookmarkIsDoneChanged(bm, new_value);
                                        }}
                                />
                            </CSSTransition>
                            )
                }
                </TransitionGroup>
            </Flex>
        </div>
    )
}

export default BookmarkList;