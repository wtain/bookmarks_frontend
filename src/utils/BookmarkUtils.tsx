
import React from "react";
import BookmarkDto from "../domain/dto/BookmarkDto";
import {v4 as uuidv4} from 'uuid';
import TagDto from "../domain/dto/TagDto";

const createBookmark = (summary: string, contents: string, tags: TagDto[]): BookmarkDto => {
    return {
                summary, 
                contents, 
                id: uuidv4(), 
                created: new Date().toLocaleString(),
                tags,
                isDone: false
            }
}

const createNewTag = (): TagDto => {
    return { name: "", id: uuidv4() }
}

const BookmarkUtils = { createBookmark, createNewTag };

export default BookmarkUtils;