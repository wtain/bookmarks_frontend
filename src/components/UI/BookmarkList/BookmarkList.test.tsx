import renderer from 'react-test-renderer';
import { BrowserRouter } from 'react-router-dom';
import BookmarkList from './BookmarkList';
import TimeAgo from 'javascript-time-ago';
import en from 'javascript-time-ago/locale/en.json';
import BookmarkDto from '../../../domain/dto/BookmarkDto';
import { AppProvider } from "@shopify/polaris";
import translations from "@shopify/polaris/locales/en.json";
import { jestHacks } from '../../../utils/JestUtils';

TimeAgo.addDefaultLocale(en);
jest
  .useFakeTimers()
  .setSystemTime(new Date('2022-09-17T13:47:36.000Z'));

function run_test(bookmarks: BookmarkDto[]) {

  const tree = renderer
  .create(
    <BrowserRouter>
      <AppProvider i18n={translations}>
        <BookmarkList bookmarks={bookmarks}  />
      </AppProvider>
    </BrowserRouter>
  )
  .toJSON();
  expect(tree).toMatchSnapshot();
}

describe('BookmarkList tests', () => {
  beforeAll(() => {
    jestHacks();
  });

  it('BookmarkList renders correctly', () => {
    const bookmarks: BookmarkDto[] = [
      {
        summary: 'Tests',
        contents: 'For UI and server',
        id: '23f69d44-2080-430e-a53d-794689a78754',
        created: new Date("2022-02-06T20:37:07.000Z"),
        tags: [
          { name: 'UI', id: 'c7bd914c-d9ff-49ea-822e-092e95252330' },
          { name: 'Server', id: '59a6fb07-46b0-4e24-9a99-ab3e7a0d4509' },
          {
            name: 'Infrastructure',
            id: '542d0f21-1bfb-4eab-939e-924ca28976da'
          },
          {
            name: 'CodeQuality',
            id: '56ca383f-eff9-4b5f-a1bd-cf8db9a5d4b1'
          }
        ],
        isDone: false,
        updated: new Date("2022-02-13T16:26:45.000Z")
      },
      {
        summary: 'Bookmark - reminder',
        contents: 'Reminder/due date property for bookmark',
        id: '633a089a-1076-4904-8650-4160db0549a9',
        created: new Date("2022-02-13T16:21:02.000Z"),
        tags: [
          {
            name: 'Functionality',
            id: 'c2ea5a69-9e52-4b1e-a6ec-53393452cd35'
          }
        ],
        isDone: false
      },
      {
        summary: 'Edit tags',
        contents: 'Ability to edit tags for existing bookmarks',
        id: 'ab2809a7-de53-4494-8ece-2aeecadce254',
        created: new Date("2022-01-30T20:48:19.000Z"),
        tags: [ { name: 'Task', id: '4b68bb47-c561-4958-8682-ea84ca39773d' } ],
        isDone: true,
        updated: new Date("2022-02-13T14:13:09.000Z")
      }
    ]

    run_test(bookmarks);
  });

  it('BookmarkList renders correctlyempty bookmark list', () => {
    const bookmarks: BookmarkDto[] = [];

    run_test(bookmarks);
  });
});