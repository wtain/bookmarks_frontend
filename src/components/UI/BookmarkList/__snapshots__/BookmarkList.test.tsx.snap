// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookmarkList renders correctly 1`] = `
<div>
  <div
    className=""
    style={Object {}}
  >
    <div>
      <div
        className="bookmark"
      >
        <div
          className="Collapsible"
        >
          <span
            aria-controls="collapsible-content-1663422456000"
            aria-disabled={false}
            aria-expanded={false}
            className="Collapsible__trigger is-closed"
            id="collapsible-trigger-1663422456000"
            onClick={[Function]}
            onKeyPress={[Function]}
            role="button"
            style={null}
            tabIndex={null}
          >
            <input
              checked={false}
              className="done_checkbox"
              onChange={[Function]}
              onClick={[Function]}
              type="checkbox"
            />
            <div
              className="header_div"
              data-for="registerTip23f69d44-2080-430e-a53d-794689a78754"
              data-tip={true}
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
            >
              <span>
                <button
                  className="btn_remove"
                  onClick={[Function]}
                >
                  x
                </button>
              </span>
              <div
                className="summary_preview"
              >
                <span
                  className="summary_preview_text"
                >
                  For UI and server...
                </span>
                (
                <a
                  href="bookmark/23f69d44-2080-430e-a53d-794689a78754"
                  onClick={[Function]}
                >
                  Link
                </a>
                )
              </div>
              <div
                className="summary"
                contentEditable={true}
                dangerouslySetInnerHTML={
                  Object {
                    "__html": "Tests",
                  }
                }
                disabled={false}
                onBlur={[Function]}
                onChange={[Function]}
                onInput={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
              />
            </div>
          </span>
          <div
            aria-labelledby="collapsible-trigger-1663422456000"
            className="Collapsible__contentOuter"
            hidden={false}
            id="collapsible-content-1663422456000"
            onTransitionEnd={[Function]}
            role="region"
            style={
              Object {
                "WebkitTransition": "height 400ms linear",
                "height": 0,
                "msTransition": "height 400ms linear",
                "overflow": "hidden",
                "transition": "height 400ms linear",
              }
            }
          >
            <div
              className="Collapsible__contentInner"
            >
              <div
                className="created"
              >
                Created: 
                7 months ago
              </div>
              <div
                className="created"
              >
                Updated: 
                7 months ago
              </div>
              <div
                className="contents"
                contentEditable={true}
                dangerouslySetInnerHTML={
                  Object {
                    "__html": "For UI and server",
                  }
                }
                disabled={false}
                onBlur={[Function]}
                onChange={[Function]}
                onInput={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
              />
              <br />
              <div
                className="container tags_container"
                onClick={[Function]}
              >
                <span
                  className="tag"
                  onClick={[Function]}
                  onMouseEnter={[Function]}
                  onMouseLeave={[Function]}
                >
                  UI
                </span>
                <span
                  className="tag"
                  onClick={[Function]}
                  onMouseEnter={[Function]}
                  onMouseLeave={[Function]}
                >
                  Server
                </span>
                <span
                  className="tag"
                  onClick={[Function]}
                  onMouseEnter={[Function]}
                  onMouseLeave={[Function]}
                >
                  Infrastructure
                </span>
                <span
                  className="tag"
                  onClick={[Function]}
                  onMouseEnter={[Function]}
                  onMouseLeave={[Function]}
                >
                  CodeQuality
                </span>
                <span
                  className="sp"
                >
                  <input
                    className="input"
                    onBlur={[Function]}
                    onChange={[Function]}
                    onKeyDown={[Function]}
                    placeholder="Tags"
                    value=""
                  />
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        className="bookmark"
      >
        <div
          className="Collapsible"
        >
          <span
            aria-controls="collapsible-content-1663422456000"
            aria-disabled={false}
            aria-expanded={false}
            className="Collapsible__trigger is-closed"
            id="collapsible-trigger-1663422456000"
            onClick={[Function]}
            onKeyPress={[Function]}
            role="button"
            style={null}
            tabIndex={null}
          >
            <input
              checked={false}
              className="done_checkbox"
              onChange={[Function]}
              onClick={[Function]}
              type="checkbox"
            />
            <div
              className="header_div"
              data-for="registerTip633a089a-1076-4904-8650-4160db0549a9"
              data-tip={true}
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
            >
              <span>
                <button
                  className="btn_remove"
                  onClick={[Function]}
                >
                  x
                </button>
              </span>
              <div
                className="summary_preview"
              >
                <span
                  className="summary_preview_text"
                >
                  Reminder/due date property for bookmark...
                </span>
                (
                <a
                  href="bookmark/633a089a-1076-4904-8650-4160db0549a9"
                  onClick={[Function]}
                >
                  Link
                </a>
                )
              </div>
              <div
                className="summary"
                contentEditable={true}
                dangerouslySetInnerHTML={
                  Object {
                    "__html": "Bookmark - reminder",
                  }
                }
                disabled={false}
                onBlur={[Function]}
                onChange={[Function]}
                onInput={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
              />
            </div>
          </span>
          <div
            aria-labelledby="collapsible-trigger-1663422456000"
            className="Collapsible__contentOuter"
            hidden={false}
            id="collapsible-content-1663422456000"
            onTransitionEnd={[Function]}
            role="region"
            style={
              Object {
                "WebkitTransition": "height 400ms linear",
                "height": 0,
                "msTransition": "height 400ms linear",
                "overflow": "hidden",
                "transition": "height 400ms linear",
              }
            }
          >
            <div
              className="Collapsible__contentInner"
            >
              <div
                className="created"
              >
                Created: 
                7 months ago
              </div>
              <div
                className="contents"
                contentEditable={true}
                dangerouslySetInnerHTML={
                  Object {
                    "__html": "Reminder/due date property for bookmark",
                  }
                }
                disabled={false}
                onBlur={[Function]}
                onChange={[Function]}
                onInput={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
              />
              <br />
              <div
                className="container tags_container"
                onClick={[Function]}
              >
                <span
                  className="tag"
                  onClick={[Function]}
                  onMouseEnter={[Function]}
                  onMouseLeave={[Function]}
                >
                  Functionality
                </span>
                <span
                  className="sp"
                >
                  <input
                    className="input"
                    onBlur={[Function]}
                    onChange={[Function]}
                    onKeyDown={[Function]}
                    placeholder="Tags"
                    value=""
                  />
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        className="bookmark"
      >
        <div
          className="Collapsible"
        >
          <span
            aria-controls="collapsible-content-1663422456000"
            aria-disabled={false}
            aria-expanded={false}
            className="Collapsible__trigger is-closed"
            id="collapsible-trigger-1663422456000"
            onClick={[Function]}
            onKeyPress={[Function]}
            role="button"
            style={null}
            tabIndex={null}
          >
            <input
              checked={true}
              className="done_checkbox"
              onChange={[Function]}
              onClick={[Function]}
              type="checkbox"
            />
            <div
              className="header_div"
              data-for="registerTipab2809a7-de53-4494-8ece-2aeecadce254"
              data-tip={true}
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
            >
              <span>
                <button
                  className="btn_remove"
                  onClick={[Function]}
                >
                  x
                </button>
              </span>
              <div
                className="summary_preview"
              >
                <span
                  className="summary_preview_text"
                >
                  Ability to edit tags for existing bookmarks...
                </span>
                (
                <a
                  href="bookmark/ab2809a7-de53-4494-8ece-2aeecadce254"
                  onClick={[Function]}
                >
                  Link
                </a>
                )
              </div>
              <div
                className="summary done"
                contentEditable={true}
                dangerouslySetInnerHTML={
                  Object {
                    "__html": "Edit tags",
                  }
                }
                disabled={false}
                onBlur={[Function]}
                onChange={[Function]}
                onInput={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
              />
            </div>
          </span>
          <div
            aria-labelledby="collapsible-trigger-1663422456000"
            className="Collapsible__contentOuter"
            hidden={false}
            id="collapsible-content-1663422456000"
            onTransitionEnd={[Function]}
            role="region"
            style={
              Object {
                "WebkitTransition": "height 400ms linear",
                "height": 0,
                "msTransition": "height 400ms linear",
                "overflow": "hidden",
                "transition": "height 400ms linear",
              }
            }
          >
            <div
              className="Collapsible__contentInner"
            >
              <div
                className="created"
              >
                Created: 
                8 months ago
              </div>
              <div
                className="created"
              >
                Updated: 
                7 months ago
              </div>
              <div
                className="contents"
                contentEditable={true}
                dangerouslySetInnerHTML={
                  Object {
                    "__html": "Ability to edit tags for existing bookmarks",
                  }
                }
                disabled={false}
                onBlur={[Function]}
                onChange={[Function]}
                onInput={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
              />
              <br />
              <div
                className="container tags_container"
                onClick={[Function]}
              >
                <span
                  className="tag"
                  onClick={[Function]}
                  onMouseEnter={[Function]}
                  onMouseLeave={[Function]}
                >
                  Task
                </span>
                <span
                  className="sp"
                >
                  <input
                    className="input"
                    onBlur={[Function]}
                    onChange={[Function]}
                    onKeyDown={[Function]}
                    placeholder="Tags"
                    value=""
                  />
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`BookmarkList renders correctlyempty bookmark list 1`] = `
<div>
  (No bookmarks)
</div>
`;
